Evaluator

ola13 opened the pull request:
This is the first stab at implementing `Evaluator`.

On top of `compute`, I implement 2 functions for statistical significance. 
- `get_confidence_interval()`
- `get_bootstrap_p_value()`
 
I follow @TristanThrush's [colab](https://colab.research.google.com/drive/198QrZx8wGpouui1X0Grl7Lt9Uogvo1_h) for confidence interval and p-value calculations (note points about speed in the notebook - this is not addressed in the current variant of the PR)

Alternatively, statsig calculations could be triggered via kwargs passed to `compute` - please let me know which variant of the API would be more intuitive.

As pointed out by @fxmarty in https://github.com/huggingface/evaluate/issues/23#issuecomment-1132821999, the current API won't work with all `pipeline` variants and I only tested on `text-classification`. I'd recommend going with a limited functionality for the release on Friday and adding coverage at a later date.

Please let me know if you feel this is moving in the right direction.

TODOs:
- finalize tests (currently only for debugging)
- add documentation
- add relevant packages to `setup.py`

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/evaluate/pr_62). All of your documentation changes will be reflected on that endpoint.

douwekiela commented:
Phew glad we recovered this! :) So if I understand Tristan's notebook correctly there are two methods: one is for computing the confidence intervals of a single classifier, which would probably go inside this `Evaluator`, and the other is to compute the bootstrap p-value of a comparison between two sets of predictions, which would probably go inside the comparison/ directory (similar to eg [mcnemar](https://github.com/huggingface/evaluate/blob/main/comparisons/mcnemar/mcnemar.py)). I may be misunderstanding something, but it seems like we mixed these up? cc @TristanThrush

ola13 commented:
Hi @douwekiela, yes this is what's happening, I can move the get_bootstrap_p_value() code there for now (or remove?).

But it might actually be useful to have API for comparing against another model within Evaluator?

ola13 commented:
Regarding uncertainty estimation via e.g. variational dropout (as mentioned in the original issue), I'll need to research the topic a bit more to propose an implementation, but open to pointers suggestions!

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -0,0 +1,106 @@
+class Evaluator:
+    def __init__(
+        self,
+        pipe: Union[str, Pipeline],
```
This should also include a `Callable` type for users that setup their own pipeline (e.g. for a `sklearn` model).

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -0,0 +1,106 @@
+        col_mapping: Optional[Dict[str, str]] = None,
+        label_mapping: Optional[Dict[str, Number]] = None,
+    ):
+        self.pipe = pipeline(pipe) if isinstance(pipe, str) else pipe
```
See my remark in the main comment regarding specialised `Evaluator` classes per task. If we have a `TextClassificationEvaluator` it would not make sense to pass another pipeline type with string, instead we could pass a model:
```suggestion
        self.pipe = pipeline(self.task, model=pipe) if isinstance(pipe, str) else pipe
```

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -0,0 +1,106 @@
+        if col_mapping is not None:
+            self.data = data.rename_columns(col_mapping)
+
+    def get_predictions(
```
```suggestion
    def _compute_predictions()
```

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -0,0 +1,106 @@
+            for element in predictions
+        ]
+
+    def get_references(self):
```
```suggestion
    def _references(self):
```

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -0,0 +1,106 @@
+    def get_references(self):
+        return self.data["references"]
+
+    def statistic(self, predictions=None, references=None, metric_key: Optional[str] = None):
```
```suggestion
    def _compute_metric(self, predictions=None, references=None, metric_key: Optional[str] = None):
```

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -0,0 +1,106 @@
+            self.metric_key if metric_key is None else metric_key
+        ]
+
+    def get_confidence_interval(self, confidence_level: float = 0.95, n_resamples: int = 9999):
```
```suggestion
    def _compute_metric_with_bootstrap(self, confidence_level: float = 0.95, n_resamples: int = 9999):
```

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -0,0 +1,106 @@
+                s += 1
+        return s / n_resamples
+
+    def compute(self):
+        predictions = self.get_predictions()
+        references = self.get_references()
+        result = self.metric.compute(predictions=predictions, references=references)
+        return result
```
```suggestion
    def compute(self, strategy="simple"):
        predictions = self.get_predictions()
        references = self.get_references()
        if strategy == "simple":
            result = self._compute_metric(predictions=predictions, references=references)
        elif strategy == "bootstrap":
            result = self._compute_metric_with_bootstrap(predictions=predictions, references=references)
        return result
```

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -0,0 +1,106 @@
+        data: Union[str, DatasetDict, Dataset, IterableDatasetDict, IterableDataset],
+        metric: Union[str, Metric],
+        metric_key: Optional[str] = None,
+        col_mapping: Optional[Dict[str, str]] = None,
+        label_mapping: Optional[Dict[str, Number]] = None,
```
What's missing a bit is some useful errors when the mappings are missing but needed or when they are wrong. We could catch the resulting errors throughout the class and provide a useful error message to the user, e.g.:
```
f"The dataset and model inputs are not matching: the dataset contains the columns {self.data.columns} but a column XXX is required to serve as model inputs. You can provide a mapping with `col_mapping`"
```

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -0,0 +1,106 @@
+        data: Union[str, DatasetDict, Dataset, IterableDatasetDict, IterableDataset],
+        metric: Union[str, Metric],
+        metric_key: Optional[str] = None,
+        col_mapping: Optional[Dict[str, str]] = None,
```
```suggestion
        column_mapping: Optional[Dict[str, str]] = None,
```

lvwerra reviewed (commented):
Hi @ola13, thanks so much for working on this! 

First, some high-level feedback. I agree that we can't have a single Evaluator for all types of tasks, so I propose that we setup the evaluator somewhat similar to the `Pipeline`: we can have a class `Evaluator` as en entry and have subclasses for each task e.g. `TextClassificationEvaluator`.

```Python
Evalutor("text-classification")
>>> TextClassificationEvaluator object
```

Independent of that I would suggest moving the `pipe`, `data`, and `metric` to the `compute` function. I think one common use-case could be:

```Python
for model in models:
    result = evaluator.compute(model, data, metric)
```

In the current implementation that would always require two steps:
```Python
for model in models:
    evaluator = Evaluator(model, data, metric)
    result = evaluator.compute()
```

For the bootstrapping I was thinking we could maybe have it as an kwarg. Then we would only need to really expose one method to the user that's called `compute`:

```Python
evaluator.compute(model, data, metric, strategy="bootstrap")
```

Regarding the `get_bootstrap_p_value`: In the library we have the distinction between `metric` (evaluates models), `comparison` (compares models), and `measurement` (evaluates data). This falls into the `comparison` category and maybe we can create a dedicated evaluator in the future `Evaluator("comparison")` for that purpose which would take two pipelines as input.

What do you think?

ola13 commented:
@lvwerra thanks for the through review! sounds good I'll iterate on the implementation in line with your suggestions :)

ola13 commented:
@lvwerra, looking again I wonder about this suggestion:
> moving the pipe, data, and metric to the compute

the function signature would actually have to be a bit uglier then that and account for `col_mapping` and `label_mapping` as well as what I call a `metric_key` (metrics return a dictionary of values, e.g. bleu, and for bootstrap calculations we need a single scalar to work with).

More generally, if we assume `model`, `data` and `metric` can all be variable, then I wonder what the point is of having evaluator as a class - wouldn't it be more intuitive to expand `compute` in the `EvaluationModule`. It would still work in the usecase you mention:
```
for model in models:
    result = metric.compute(model, data, col_mapping=col_mapping, label_mapping=label_mapping, strategy="bootstrap")
```


If I were to make a case for the existence of `Evaluator`, I'd say it should encapsulate a dataset (data mappings would be done in initialisation) and a metric, with the model along with label mapping passed to `compute` - this would serve the usecase of comparing multiple models against a unique (data, metric) combination which seems like a common need.

What do you think? :)

lvwerra commented:
Hi @ola13 

Thanks for looking more into this! I don't think we should keep the dataset fixed and allow the model to be changed. I think the other use-case of having one model you want to evaluate on several datasets is as common. In the future maybe even with several metrics which we could enable with #8. So for the use-case with variable datasets and models:

```Python
# version 1
evaluator = Evaluator("text-classification")

for data in datasets:
    for model in models:
        result = evaluator.compute(model, data, metric, **kwargs)

# version 2
for data in datasets:
    for model in models:
        evaluator = Evaluator("text-classification", data, metric, **kwargs)
        result = evaluator.compute(model)
```

I see your point that the `compute` gets more polluted but I don't think having it static at initialization simplifies these use-cases much.

Adding more flexibility to `EvaluationModule` would require a massive refactor as it is heavily built around the idea of having only data as inputs and not doing any inference. I think it is easier for the user if a the `EvaluationModule` always takes references and predictions and not also inputs, models and predictions. 

Finally on your point regarding the `metric_key `: I see two options to resolve this without needing an explicit key to be passed:
1. Most metrics have the main score in a key with the same name. E.g. `accuracy` returns `{"accuracy": 1.0}`. We could check if that key is available.
2. Couldn't we apply the bootstrapping for each key in the dictionary? I think that would be the best option if possible.
3. (coming soon): we are thinking about adding a postprocessor (#50) which won't be ready for release but could also take care of this. 

What do you think?

lvwerra commented on src/evaluate/__init__.py:
```diff
@@ -26,11 +26,54 @@
 from .utils import *
 from .utils import gradio, logging
+
+
```
I would put these helper variables and functions in the evaluator.py module to keep the __init__.py lean.

lvwerra commented on src/evaluate/__init__.py:
```diff
@@ -26,11 +26,54 @@
+
+SUPPORTED_EVALUATOR_TASKS = {
+    "text-classification": {
+        "impl": TextClassificationEvaluator,
```
```suggestion
        "implementation": TextClassificationEvaluator,
```

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -12,95 +12,125 @@
-
-    def get_predictions(
-        self, baseline_pipe: Optional[Pipeline] = None, baseline_label_mapping: Optional[Dict[str, Number]] = None
+        model: str = None,
+        pipe: Union[Pipeline, Callable] = None,
```
Having two arguments for the model/pipeline is not ideal I think: It makes it a bit less clear for a user what needs to be provided (a model, a pipe, or both?). What do you think about having it 
```suggestion
        model_or_pipeline: Union[Pipeline, Callable, str] = None,
```

Not that in the `transformers` ecosystem, especially around the pipelines users can usually pass a model instance or a string as model. So we should check if a user passed a model instance.

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -12,95 +12,125 @@
+        label_column: str = "references",
+        label_mapping: Optional[Dict[str, Number]] = None,
+    ) -> Tuple[Dict[str, float], Any]:
+        assert data is not None
```
Could we add Errors instead of asserts with messages that guide the user to fix the problem?

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -12,95 +12,125 @@
+    ) -> Tuple[Dict[str, float], Any]:
+        assert data is not None
+
+        pipe = (
+            pipeline(self.task, model=model)
+            if model is not None
+            else pipe
+            if pipe is not None
+            else pipeline(self.task)
```
Could we replace this with normal if/else clauses for readability?

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -12,95 +12,125 @@
-                baseline_preds[resampled_indices], references[resampled_indices]
+        assert pipe.task == self.task
+
+        metric = (
+            load_metric(metric)
+            if isinstance(metric, str)
+            else (metric if metric is not None else (self.default_metric if self.default_metric is not None else None))
+        )
```
Same as above.

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -12,95 +12,125 @@
-import numpy as np
-from datasets import Dataset, DatasetDict, IterableDataset, IterableDatasetDict, load_dataset, load_metric
+from datasets import Dataset, load_dataset, load_metric
 from datasets.metric import Metric
```
`evaluate` is replacing `datasets.load_metric` and `Metric`. For `load_metric` you can use `evaluate.load` and the new class for `Metric` is `EvaluationModule`. Sorry, didn't catch this earlier!

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -12,95 +12,125 @@
+        metric = (
+            load_metric(metric)
+            if isinstance(metric, str)
+            else (metric if metric is not None else (self.default_metric if self.default_metric is not None else None))
```
`self.default_metric` is also a string, right? This would fail in that case as `metric` would end up as the default string.

lvwerra reviewed (commented):
Thanks @ola13, that's exciting progress. I left a few comments, overall I think we are converging towards a great solution!

douwekiela commented:
> Regarding uncertainty estimation via e.g. variational dropout (as mentioned in the original issue), I'll need to research the topic a bit more to propose an implementation, but open to pointers suggestions!

Regarding variational dropout for uncertainty estimation @ola13, this is probably the best starting point: http://mlg.eng.cam.ac.uk/yarin/blog_3d801aa532c1ce.html

ola13 commented:
@lvwerra I've updated the core functionality as per your suggestions, let me know if it looks good!

Some considerations around `model_or_pipeline` - I treat a model type or a string argument as a model, and a pipeline / callable argument as a pipeline. In order to make it work when models are provided I also added an Optional tokenizer parameter to `compute()`, which is ignored if  `model_or_pipeline` is indeed a pipeline - let me know if that's user friendly!

If overall logic is fine, I can finalize tests / documentation.

@douwekiela - thanks for the pointer! I'll read up - I think we can add a variational dropout functionality to the base `Evaluator` function similarly to how `_compute_confidence_interval()` is done right now - I'd leave that until post-launch though.

Currently the `Evaluator` class encapsulates a `task` and a `default_metric` - it is best suited for evaluating a (model, data) combination against a single metric. We can discuss how we can extend the API to perform comparisons of two (or more?) models on a dataset using a metric and statistic significance functionalities from the `comparisons` module - but I'd recommend leaving it until post-launch. 

Let me  know what you think!

lvwerra commented on src/evaluate/evaluator.py:
```diff
@@ -88,7 +94,8 @@ def _compute_predictions(self, pipe: Pipeline, inputs, label_mapping: Dict[str,
         self,
-        model_or_pipeline=None,
+        model_or_pipeline: Union[str, Pipeline, Callable, PreTrainedModel, TFPreTrainedModel] = None,
+        tokenizer: Optional[Union[str, PreTrainedTokenizer]] = None,
```
Just a nit: can we move the tokenizer after data/metric. I think that conveys a bit better that in most cases you just need to provide model/data/metric and not a tokenizer.

ola13 replied:
sounds good, basically on a small example I tested the tokenizer wasn't compatible, that's why I put it here, but happy to remove for clarity.

lvwerra replied:
I mean we can keep it but just change the orderings of the arguments :)

lvwerra reviewed (commented):
Your points sound good to me, I agree that we can add these other features after the initial launch! Are you happy to work on the docstrings and tests? I can help work on documentation after we merge.

ola13 commented:
Pushed tests, however they are very slow, I'll try to mock predictions computations instead of making a forward pass on models

lvwerra reviewed (approved):
LGTM! 🚀
